### ISSUE TITLE
Adding Backwards Edge in Tensorflow Java

### ISSUE THREAD

**Issue by samdow** [tensorflow/tensorflow#24231]
**System information**
- TensorFlow version (you are using): 1.12
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
As we started working with Tensorflow Java, we realized that it isn't yet possible to implement while loops in the graph because we cannot mutate the nodes after they have been created. Therefore, we cannot make a backwards edge to create the loop from the NextIteration node to the Merge node, as seen in [the design of while loops](https://github.com/tensorflow/tensorflow/blob/2a050766bf0556d7d92eea62d40fd2bebbcb399f/tensorflow/cc/ops/while_loop.cc#L148).

**Will this change the current api? How?**
Yes. This will allow the graph to be mutated after a node has been created.

**Who will benefit with this feature?**
Those who want to use Tensorflow Java to make a while loop or other sorts of loops.

**Any Other info.**

**Comment by asimshankar** [comment]
(Apologies for the delayed response, didn't notice this issue earlier).

Backward edges are a bit iffy and making the API more prominent makes it easy to create invalid/incorrect graphs, which I'd like to avoid.

Even in Python we are moving towards a representation of loops in the graph that does not involve backward edges (see https://github.com/tensorflow/tensorflow/blob/1d1aeb64b91effd4329965cc7954a3e21820c88b/tensorflow/python/ops/control_flow_util.py#L31 and the [RFC for functional while loops](https://github.com/tensorflow/community/blob/master/rfcs/20180821-differentiable-functional-while.md)).

If implementing while loops in Java, I'd recommend that we pursue this functional while loop approach instead. Thoughts?

**Comment by irenedea** [comment]
Thanks for the information @asimshankar! 

Reading through the RFC, it seems that there is support for taking the gradient of the entire while loop; however, we were wanting to calculate gradients *within* the body of the while loop (for example, training a model via gradient descent). Are there any plans to add support for this use case?

**Comment by asimshankar** [comment]
I'm probably missing something, but why do edges need to be updated when computing the gradient within a single iteration of the loop? The original issue description talks about creating an edge from a `NextIteration` to a `Merge`, but these don't appear within the body subgraph of the loop, right?

Could you describe the graph that you want to create in a bit more detail? Thanks.

**Comment by irenedea** [comment]
You are correct that we do not need to use `updateEdge` for anything other than creating an edge from a `NextIteration` to a `Merge`. This issue description was purely to add support for creating a while loop.

The functional solution outlined in the RFC looks promising, but from playing around with the python `while_v2` implementation, it errors when we attempt to  generate gradients within the body. We understand this is a work in progress, so just wanted to ask: will it be possible to include `addGradients` (or rather, the subgraph generated by `addGradients`) in the body subgraph in order to perform gradient descent?

**Comment by skye** [comment]
Hi @samdow, sorry for missing this earlier! Please assign or mention me in any future while_loop issues you file.

One potential problem for Tensorflow Java is that while_v2 gradients are only implemented in Python, and not in the C API. So the design will have to be ported to the C API, or however Tensorflow Java is computing gradients (unless it's somehow calling the Python code).
